### DISCUSSION TITLE
EDDSA Public Key Verification

### DISCUSSION THREAD

**Discussion by abkarcher** [openssl/openssl#23380] [Q&A] [answered]
Hello!

I am attempting to perform ACVP testing on OpenSSL's EDDSA implementation. While is it not currently included in the FIPS provider, it seems likely to be in the future at some point.

While I have key generation, signature generation, and signature verification working, I am unable to get key verification working.  The ACVP test provides a simple raw "q"/public key value and a curve, and expects the implementation to specify whether the key is valid for that curve.

The function seemingly used here for checking keys, "ecx_validate", is not able to verify a public key by itself (it only checks to make sure it exists, and returns ok if so). This would be needed to include EDDSA key verification in the FIPS provider in the future. It also raises a more general concern - should the EVP_PKEY_public_check API (and other calls that eventually try to validate the key) return that a public key is valid when it hasn't been checked? It will return a success for any key value so long as it is an okay length. This seems to apply to both ED25519 and ED448.

Is there any plan to add this functionality soon? or is there any other APIs I could use to achieve similar functionality?

Thanks,
Andrew

**Comment by t8m** [comment, marked as answer]
There is not much point in doing a public key verification for EdDSA algorithms as the public key is encoded in the compressed form and checking whether the point is on curve happens naturally as part of the decompression which happens in the signature verification algorithm. So a verification with invalid public key will fail.

> Should the EVP_PKEY_public_check API (and other calls that eventually try to validate the key) return that a public key is valid when it hasn't been checked?

Yes, otherwise applications that use the call in case they support multiple types of public key algorithms would have to explicitly skip the call for EdDSA public keys.

Is it actually required for FIPS validation that a public key verification is provided by the library? I think it is not if it is built-in the signature verification.

**Comment by abkarcher** [comment]
Thanks for the response!

Trying to find out on my end if key verification is in fact required. It isn't explicitly mentioned in FIPS 186-5 either. Will report here for thoroughness' sake if I find anything out.

Thanks again,
Andrew

**Reply by slontis** [reply]
I am looking into putting EDDSA (but not X25519) into the FIPS boundary. Did you check with a lab if the KeyVer is required?

**Reply by slontis** [reply]
I assume it is not required since this seems like a KAS requirement (which is not standardized).

**Reply by abkarcher** [reply]
I have not been able to get any solid information about this one way or the other, unfortunately. All I can add is, anecdotally, I have not seen many general-purpose modules testing signature schemes (typically key generation, key verification, signature generation, and signature verification) test some of those modes of operation but not the others.

On a side note, I do have an ACVP testing harness for EDDSA in the latest versions of libacvp here - https://github.com/cisco/libacvp/blob/main/app/app_eddsa.c if that helps with any effort to move it into the FIPS boundary. It is disabled currently, but everything except keyver should work. If there is anything we can do to assist with this, feel free to reach out.

**Reply by slontis** [reply]
I have added KeyVer in PR #25265.
 If that is approved you should be able to do all ED FIPS tests against master

**Comment by slontis** [comment]
Are you aware of anything else that is not working? We are trying to get master in a state where another FIPS 140-3 validation could be done.

**Reply by abkarcher** [reply]
The other thing I ran into recently was #24051 which seems to have been addressed. I can do some ACVP runs against the new PR and master to see.

The most recent algorithm cert I see at NIST is for the 3.1.2 FIPS provider. Is there any other new algorithms that would be tested/have been added to FIPS since then other than RSA Decryption Primitive and EDDSA?

**Reply by slontis** [reply]
The main changes we have been doing relate to enforcing FIPS approved mode (using FIPS indicators).
This means some algorithms are no longer allowed as FIPS approved.
(See issue #24849)

e.g DSA keygen, signing (DSA verify still allowed),
TDES encryption
EC keygen enforces >= 112 bits.
RSA PKCS15 Encryption, RSA X931 signatures are not approved,
Enforcing Allowed Digests for all algorithms, and doing minimum keysize and output size checks
RSA pss salt length restrictions enforced.

**Reply by abkarcher** [reply]
Thanks. I will try to get a good algorithm capabilities registration together and run some tests early-mid next week.

**Reply by abkarcher** [reply]
Just an update, still working on this  - I can confirm EDDSA and RSA decryption primitive are passing all ACVP tests, and the FIPS 186-5 tests for RSA and ECDSA pass as well. Delayed slightly trying to sort out new minimum key sizes for things, but I should have all supported algorithms done in the next day or two and will provide the registration used and logs. One question - on the 3.1.2 algorithm cert, hash DRBG and hmac DRBG test SHA3 256 and 512, but not SHA2 256 and 512, which are also allowed - was this intentional, or should those be tested as well?

**Reply by paulidale** [reply]
That's definitely a question for the lab to answer.  I assumed all flavours would be tested.  There was some back and forth over  some digests usages being missed in various places, it's likely this slipped through the cracks.

There should still be time to run the tests and update the paperwork.  FYI for @nhorman

**Reply by abkarcher** [reply]
Hello,

I have completed testing. Apologies in advance for the long-winded message; just want to provide as much info as possible. Everything I have tested seems to work fine, except for one thing I am not clear on. Attempting to perform KAS-ECC testing with B or K curves returns an error:

```
40F7D835D77F0000:error:1C8000EC:Provider routines:ecdh_plain_derive:cofactor required:providers/implementations/exchange/ecdh_exch.c:554:

```

I am not sure if FIPS changes have intentionally affected the availability of these curves, but it is the same code used to test 3.1.2, and the curves still run for ECDSA. Regardless, I performed the rest of the testing with those curves disabled. I have included both the capabilities registration used, and the log of the run. 

[ssl_340_registration_acvp.json](https://github.com/user-attachments/files/16883508/ssl_340_registration_acvp.json)
[logs.txt](https://github.com/user-attachments/files/16883509/logs.txt)

  Finally, here is a list of the general changes I made for the registration compared to 3.1.2 that I can tell:
-  Added EDDSA (all four modes)
-  Removed X931 RSA signatures
-  Moved RSA revision to 186-5
    -  Removed all SHA1 abilities from verify (Required, testing not supported)
    -  Removed all 1024 length support from verify (Required, testing not supported)
    -  Specified MGF1 as mask function (Assuming SHAKE not currently supported as mask function)
-  Moved ECDSA revision to 186-5
    -  Removed all SHA1 abilities from verify (Required)
    -  Removed B-163, K-163, and P-192 curves from verify (Required, testing not supported)
-  New RSA Signature Primitive revision - support 2048/3072/4096 modulo testing; standard keys as well as CRT keys
-  Add RSA Decryption Primitive testing - same capabilities as sigprim
-  Removed DSA generation testing, left verify in
-  HMAC - set minimum key size to 112 
-  Added SHA1, SHA 2 256 and 512 to hash and hmac drbg
-  Unsure: removed B and K curves for KAS-ECC testing due to aforementioned issue


  If there are any changes that should be made to the registration, any other tests I can run, etc. please let me know, I am happy to help. I do apologize for my delay here. 
  
  Thanks!
  Andrew

**Reply by slontis** [reply]
KAS requires ECC CDH to be used (The 3rd C is cofactor).

For the B & K related curves, are you setting the  "ecdh-cofactor-mode" parameter?
These are the only NIST ECC FIPS curves that have a cofactor that is not one, and KAS requires this to be set.
i.e. Inside ossL_ecdh_simple_compute_key()
point   = cofactor * host_private_key *peer_public_key

Without the flag the cofactor is not used. (And it doesnt matter on normal curves because it is 1)

**Reply by abkarcher** [reply]
I have been setting OSSL_PKEY_PARAM_USE_COFACTOR_ECDH = 1 ("use-cofactor-flag") in the setup of the keys themselves previously. I will see about using ecdh-cofactor-mode in the derive step.

**Reply by abkarcher** [reply]
The error seems to be caused by using the "use-cofactor-flag" param on the pkeys in 3.4.0. If I use that flag, either by itself or with ecdh-cofactor-mode, I get the error. If I use ecdh-cofactor-mode by itself, the tests pass! Regardless, there seems to be some behavior change with "use-cofactor-flag".

Previous versions of OpenSSL pass using ecdh-cofactor-mode, so I have updated my code to follow that process for all versions. I have re-run the tests with B and K curves included and am including the new registration and log files -

[registration_fp_340.json](https://github.com/user-attachments/files/16899632/registration_fp_340.json)
[logs.txt](https://github.com/user-attachments/files/16899633/logs.txt)

The code used to test is all live at this branch: https://github.com/cisco/libacvp/tree/libacvp_2_2_0-throttle. Please note this is a development branch and may have minor imperfections until a release is done.

Thanks,
Andrew

**Comment by slontis** [comment]
Finally got around to testing the cofactor.. You are correct, it is a bug. PR #25548 should address this issue.
